src/services/hsm.service.js

```diff
@@ -0,0 +1,203 @@
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+const { hsm } = require('@dinamonetworks/hsm-dinamo');
+
+const { getConfig } = require('../settings');
+const { lock, lockTypes } = require('../lock');
+const { hsmBusyErrorMessage } = require('../helpers/constants');
+const { HsmError } = require('../errors');
+
+/**
+ * Connects to the HSM and returns the connection object
+ * @returns {Promise<Hsm>}
+ */
+async function hsmConnect() {
+  // Ensures that only one connection will be open at any given moment
+  const canStart = lock.lock(lockTypes.HSM);
+  if (!canStart) {
+    throw new HsmError(hsmBusyErrorMessage);
+  }
+
+  // Gets the connection data from the global config file
+  const config = getConfig();
+  const hsmConnectionOptions = {
+    host: config.hsmHost,
+    authUsernamePassword: {
+      username: config.hsmUsername,
+      password: config.hsmPassword,
+    },
+  };
+
+  // Connects to HSM and returns the connection object
+  return hsm.connect(hsmConnectionOptions);
+}
+
+/**
+ * Disconnects the headless wallet from the HSM.
+ * The connection object is a singleton, so it must not be informed as parameter.
+ * @returns {Promise<void>}
+ */
+async function hsmDisconnect() {
+  await hsm.disconnect();
+  lock.unlock(lockTypes.HSM);
+}
+
+/**
+ * Checks if the informed HSM key name contains a valid BIP32 xPriv key
+ * @param {Object} hsmConnection
+ * @param {string} hsmKeyName
+ * @returns {Promise<{reason?: string, isValidXpriv: boolean, hsmKeyName}>}
+ */
+async function isKeyValidXpriv(hsmConnection, hsmKeyName) {
+  const returnObject = {
+    hsmKeyName,
+    isValidXpriv: false,
+    reason: '',
+  };
+
+  // Checks if the key exists on HSM
+  let keyInfo;
+  try {
+    keyInfo = await hsmConnection.blockchain.getKeyInfo(hsmKeyName);
+  } catch (e) {
+    // Special treatment for a wallet not configured on the HSM
+    returnObject.reason = e.errorCode === 5004
+      ? `Key does not exist on HSM.`
+      : `Unexpected error on key retrieval: ${e.message}`;
+    return returnObject;
+  }
+
+  // Validates the key type and version
+  if (keyInfo.type !== hsm.enums.BLOCKCHAIN_KEYS.BIP32_XPRV) {
+    returnObject.reason = `hsm-key is not a valid xPriv.`;
+    return returnObject;
+  }
+  if (!(keyInfo.ver in [
+    hsm.enums.VERSION_OPTIONS.BIP32_HTR_MAIN_NET,
+    hsm.enums.VERSION_OPTIONS.BIP32_HTR_TEST_NET,
+  ])) {
+    returnObject.reason = `hsm-key is not a valid Hathor xPriv.`;
+    return returnObject;
+  }
+
+  // Returns a positive result for a valid xPriv key
+  delete returnObject.reason;
+  returnObject.isValidXpriv = true;
+  return returnObject;
+}
+
+/**
+ * Derivates an HTR xPriv key from a given xPriv key
+ * @param {Object} hsmConnection
+ * @param {string} hsmKeyName
+ * @param {Object} [options]
+ * @param {boolean} [options.isReadOnlyWallet] If true, derivation will stop at the account level
```

Comment:
Can you start a non read only wallet?

Reply:
For future implementations we may want to execute the fourth derivation for calculating the change addresses on the HSM. This flag allows this alternative.

Reply:
I didn't understand the reason for that, to be honest, when we want to derive until the change path and when we want to derive until the account path?

Reply:
As we've talked, I added some clarifications to the method parameters regarding which levels are derived from the inputs. Done on [2bf7108](https://github.com/HathorNetwork/hathor-wallet-headless/pull/356/commits/2bf7108313b675fabdeb32f7a4d60bdd853d86df).